src/core/Fingerprint.cpp

```diff
@@ -0,0 +1,240 @@
+#include "Fingerprint.hpp"
+#include "../helpers/Log.hpp"
+#include "src/config/ConfigManager.hpp"
+
+#include <filesystem>
+#include <unistd.h>
+#include <pwd.h>
+
+#include <cstring>
+#include <thread>
+
+static const auto FPRINT        = "net.reactivated.Fprint";
+static const auto DEVICE        = "net.reactivated.Fprint.Device";
+static const auto MANAGER       = "net.reactivated.Fprint.Manager";
+static const auto LOGIN_MANAGER = "org.freedesktop.login1.Manager";
+static const auto RETRY_MESSAGE = "Could not match fingerprint. Try again.";
+
+CFingerprint::CFingerprint() {
+    static auto* const PFINGERPRINTREADY   = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:fingerprint_ready_message"));
+    m_sFingerprintReady                    = *PFINGERPRINTREADY;
+    static auto* const PFINGERPRINTPRESENT = (Hyprlang::STRING*)(g_pConfigManager->getValuePtr("general:fingerprint_present_message"));
+    m_sFingerprintPresent                  = *PFINGERPRINTPRESENT;
+}
+
+static void fingerprintUnlockCallback(std::shared_ptr<CTimer> self, void* data) {
+    g_pHyprlock->unlock();
+}
+
+void CFingerprint::start() {
+    static auto* const PENABLEFINGERPRINT = (Hyprlang::INT* const*)g_pConfigManager->getValuePtr("general:enable_fingerprint");
+    const bool         ENABLEFINGERPRINT  = **PENABLEFINGERPRINT;
+    if (!ENABLEFINGERPRINT) {
+        return;
+    }
+    std::thread([this]() {
+        // For grace or SIGUSR1 unlocks
+        if (g_pHyprlock->isUnlocked())
+            return;
+
+        setupDBUS();
+
+        if (m_bAuthenticated)
+            g_pHyprlock->addTimer(std::chrono::milliseconds(1), fingerprintUnlockCallback, nullptr);
+    }).detach();
+}
+
+std::optional<std::string> CFingerprint::getLastMessage() {
+    return m_sDBUSState.message.empty() ? std::nullopt : std::optional(m_sDBUSState.message);
+}
+
+void CFingerprint::terminate() {
+    if (m_sDBUSState.connection)
+        m_sDBUSState.connection->leaveEventLoop();
+}
+
+bool CFingerprint::setupDBUS() {
+    m_sDBUSState.connection = sdbus::createSystemBusConnection();
+
+    registerSleepHandler();
+
+    if (m_sDBUSState.sleeping) {
+        m_sDBUSState.connection->enterEventLoop();
+    } else if (startVerify()) {
+        m_sDBUSState.message = m_sFingerprintReady;
+        g_pHyprlock->enqueueForceUpdateTimers();
+        m_sDBUSState.connection->enterEventLoop();
```

Comment:
NITPICK: would be a bit easier to understand if `enterEventLoop` would not happen in `setupDBUS`, but either in the std::thread ctor or in a dedicated `enterLoop` function.

Reply:
Cleaned this up a bit and moved the logic into the thread ctor.